File: 练习/每日一题/知道秘密的人数.cpp
#include<bits/stdc++.h>
using namespace std;

/*
dp[i][j]: 第 i 天时，已知秘密“恰好 j 天”的人数
dp[i][1] = dp[i-1][delay] + dp[i-1][delay+1] + ... + dp[i-1][forget-1]
 */
class Solution {
public:
    int peopleAwareOfSecret(int n, int delay, int forget) {
        const int MOD = 1e9 + 7;
        using int64 = long long;
        vector<vector<int>> dp(n+1,vector<int>(forget+1,0));
        dp[1][1] = 1;
        for(int i = 2; i<=n; i++)
        {
            for(int j = 2;j<=forget; j++)
            {
                dp[i][j] = dp[i-1][j-1];
            }

            int64 add = 0;

            for(int k = delay; k<=forget-1; k++)
            {
                add+=dp[i-1][k];
                if(add >= MOD) add -= MOD;
            }
            dp[i][1] = add;
        }

        int64 ans = 0;
        for(int j = 1; j<=forget; j++)
        {
            ans+=dp[n][j];
            if(ans>=MOD) ans-=MOD;
        }
        return int(ans);
    }
};

File: 练习/每日一题/知道秘密的人数.py
#直接向后推知道的人数
class Solution:
    def peopleAwareOfSecret(self, n: int, delay: int, forget: int) -> int:
        # dp[i] 第i天知道秘密的人
        mod = 10**9+7
        dp = [0] * (n+1)
        dp[1] = 1

        for i in range(1,n+1):
            for j in range(i+delay,i+forget):
                if j<= n:
                    dp[j] += dp[i]  
                    dp[j] %= mod
        
        res = 0
        for i in range(n+1-forget,n+1):
            res+=dp[i]
            res%=mod
        return res
